sound/soc/sof/sof-dai-clk.c

```diff
@@ -0,0 +1,153 @@
+
+	/* set IPC DAI clock control */
+	clkctrl.hdr.size = sizeof(clkctrl);
+	clkctrl.hdr.cmd = SOF_IPC_GLB_DAI_MSG | SOF_IPC_DAI_CLKCTRL;
+
+	clkctrl.type = dai_clk->type;
+	clkctrl.dai_index = dai_clk->dai_index;
+
+	clkctrl.clk_en = enable ? 1 : 0;
+	clkctrl.clk_id = dai_clk->clk_id;
+
+	dev_dbg(sdev->dev, "clkctrl: SSP%d clk_en %d clk_id %d\n",
+		clkctrl.dai_index, clkctrl.clk_en, clkctrl.clk_id);
+
+	/* send IPC to the DSP */
+	ret = sof_ipc_tx_message(sdev->ipc, clkctrl.hdr.cmd, &clkctrl,
+				 sizeof(clkctrl), &reply, sizeof(reply));
+
+	if (ret < 0) {
+		dev_err(sdev->dev,
+			"failed to control DAI clock for type %d index %d\n",
+			dai_clk->type, dai_clk->dai_index);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int sof_clk_prepare(struct clk_hw *hw)
+{
+	struct sof_clk *clkdev = container_of(hw, struct sof_clk, hw);
+	struct snd_sof_dev *sdev = clkdev->sdev;
+
+	return sof_clk_ctrl(sdev, clkdev->dai_clk, 1);
+}
+
+static void sof_clk_unprepare(struct clk_hw *hw)
+{
+	struct sof_clk *clkdev = container_of(hw, struct sof_clk, hw);
+	struct snd_sof_dev *sdev = clkdev->sdev;
+
+	sof_clk_ctrl(sdev, clkdev->dai_clk, 0);
+}
+
+static const struct clk_ops sof_clk_ops = {
+	.prepare = sof_clk_prepare,
+	.unprepare = sof_clk_unprepare,
+};
+
+int snd_sof_dai_clks_probe(struct snd_soc_component *component)
+{
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
+	struct snd_sof_pdata *pdata = sdev->pdata;
+	struct device *dev = component->dev;
+	int i, ret;
+
+	sdev->num_clks = pdata->num_clks;
+
+	if (sdev->num_clks > SOF_DAI_CLKS)
+		sdev->num_clks = SOF_DAI_CLKS;
+
+	for (i = 0; i < sdev->num_clks; ++i) {
+		struct clk_init_data init = {};
+		struct clk_lookup *dai_clk_lookup;
+		struct sof_clk *clkdev;
+
+		clkdev = devm_kzalloc(dev, sizeof(*clkdev), GFP_KERNEL);
+		if (!clkdev) {
+			ret = -ENOMEM;
+			goto err;
+		}
+
+		init.name = pdata->dai_clks[i].name;
+		init.ops = &sof_clk_ops;
+		init.flags = CLK_GET_RATE_NOCACHE | CLK_SET_RATE_GATE;
+		init.parent_names = NULL;
+		init.num_parents = 0;
+
+		clkdev->hw.init = &init;
+		clkdev->sdev = sdev;
+		clkdev->dai_clk = &pdata->dai_clks[i];
+
+		ret = devm_clk_hw_register(dev, &clkdev->hw);
+		if (ret) {
+			dev_warn(dev, "failed to register %s: %d\n",
+				 init.name, ret);
+			goto err;
+		}
+
+		dai_clk_lookup = clkdev_hw_create(&clkdev->hw, init.name,
+						  NULL);
+		if (!dai_clk_lookup) {
+			dev_warn(dev, "failed to create %s: %d\n",
+				 init.name, ret);
+			ret = -ENOMEM;
+			goto err;
+		}
+
+		sdev->dai_clks_lookup[i] = dai_clk_lookup;
+	}
+
+	return 0;
+
+err:
+	do {
+		if (sdev->dai_clks_lookup[i])
+			clkdev_drop(sdev->dai_clks_lookup[i]);
+	} while (i-- > 0);
+
+	return ret;
+}
+
+void snd_sof_dai_clks_remove(struct snd_soc_component *component)
+{
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
+	int i;
+
+	for (i = sdev->num_clks - 1; i >= 0; --i) {
+		if (sdev->dai_clks_lookup[i])
+			clkdev_drop(sdev->dai_clks_lookup[i]);
+	}
```

Comment:
same here: I assume, this is only called if `.probe()` above was successful. In that case you know, that all `sdev->dai_clks_lookup[0..sdev->num_clks-1]` elements have been created successfully